File: c++/src/Letscode.cpp
//============================================================================
// Name        : Letscode.cpp
// Author      : Martin Seidel
// Version     :
// Copyright   : 
// Description : Liest die JSON Definition eines Graphen ein, und findet darin den kürzesten Pfad vom Knoten Erde zum Knoten b3-r7-r4nd7
//============================================================================

#include <iostream>
#include <fstream>
#include <vector>
#include <queue>

#include "json.hpp"
using json = nlohmann::json;

#include "Planet.h"

using namespace std;

//Vergleichsfunktion für die PriorityQueue
class priority {
public:
  bool operator()(pair<letscode::Planet*, double>& p1, pair<letscode::Planet*, double>& p2) {
    return p1.second > p2.second;
  }
};

/**
 * Implementierung des Dijkstra-Algorithmus
 */
pair<double, vector<letscode::Planet*>> dijkstra(letscode::Planet* start, letscode::Planet* target) {
  priority_queue<pair<letscode::Planet*, double>, vector<pair<letscode::Planet*, double>>, priority> pq;
  vector<string> visited; //Speichert die schon abgeschlossenen Knoten
  map<string, double> costs; //Speichert die Kosten, zu denen jeder Knoten aktuell erreichbar ist.
  map<string, letscode::Planet*> predecessor; //Speichert zu jedem Knoten den Vorgänger, um nacher den Pfad rekonstruieren zu können.

  //init
  predecessor.emplace(start->getLabel(), static_cast<letscode::Planet*>(NULL));
  pq.push(make_pair(start, 0.0));
  double result = -1.0;

  //Hauptschleife
  while (!pq.empty()) {
    pair<letscode::Planet*, double> cur = pq.top();
    pq.pop();

    if (cur.first->getLabel() == target->getLabel()) {
      //Ziel gefunden!
      result = cur.second;
      break;
    }

    if (find(visited.begin(), visited.end(), cur.first->getLabel()) != visited.end()) {
      //Schon besuchte Knoten müssen nicht nochmal untersucht werden.
      continue;
    }
    visited.push_back(cur.first->getLabel());

    for (pair<letscode::Planet* ,double> next : cur.first->getRoutes()) {
      if (find(visited.begin(), visited.end(), next.first->getLabel()) != visited.end()) {
        //Schon besuchte Knoten müssen nicht nochmal untersucht werden.
        continue;
      }

      auto cost = costs.find(next.first->getLabel());
      double next_cost = cur.second + next.second;
      if (cost == costs.end()) {
        costs.emplace(next.first->getLabel(), next_cost);
        predecessor.emplace(next.first->getLabel(), cur.first);
        pq.push(make_pair(next.first, next_cost));
      } else if(next_cost < cost->second) {
        costs.at(next.first->getLabel()) = next_cost;
        predecessor.at(next.first->getLabel()) = cur.first;
        pq.push(make_pair(next.first, next_cost));
      }

    }
  }

  vector<letscode::Planet*> ret;

  if (result != -1.0) {
    while (target != NULL) {
      ret.push_back(target);
      target = predecessor.at(target->getLabel());
    }
    reverse(ret.begin(), ret.end());
  }

  return make_pair(result, ret);
}


int main() {

  //Einlesen des Json Files
	ifstream is("generatedGraph.json");
	json j;
	is >> j;

	//Aufbau einer Datenstruktur für den Inhalt
	vector<letscode::Planet> graph;

	for (auto p : j["nodes"]) {
	  graph.push_back(letscode::Planet(p["label"]));
	}
	for (auto r : j["edges"]) {
	  graph.at(r["source"]).addRoute(&graph.at(r["target"]), r["cost"]);
	  graph.at(r["target"]).addRoute(&graph.at(r["source"]), r["cost"]);
	}

	//Start und Zielknoten finden
	auto e = find(graph.begin(), graph.end(), "Erde");
	auto b = find(graph.begin(), graph.end(), "b3-r7-r4nd7");

	if (e == graph.end() || b == graph.end()) {
	  cerr << "Fehler: Planet Erde oder b3-r7-r4nd7 nicht gefunden!" << endl;
	  return -1;
	}

	letscode::Planet* erde = &(*e);
	letscode::Planet* bertrandt = &(*b);

	//Berechung des kürzesten Pfades mit dem Dijkstra Algorithmus
	auto result = dijkstra(erde, bertrandt);
	cout << "Lösung:" << endl;
	cout << "Gesamtentfernung: " << result.first << endl;
	cout << "Pfad: " << *result.second.at(0);
	for (size_t i = 1; i < result.second.size(); ++i) {
	  cout << " -> " << *result.second.at(i);
	}
	cout << endl;

	return 0;
}

File: python/letscode.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
'''
Created on 03.05.2019

@author: Martin Seidel
'''

import heapq
import json
import os
import sys

#Basisklasse der Datenstruktur: Symbolisiert einen Knoten
#Enthält eine Liste mit den ausgehenden Kanten, die als Tupel (Ziel, Kosten) gespeichert werden.
#Speichert außerdem die Kosten, die der Dijkstra-Algorithmus berechnet.
class Planet():

  def __init__(self, _id, name):
    self._id = _id
    self.name = name
    self.routes = []
    self.cost = float('inf')

  def addRoute(self, target, cost):
    if (target, cost) not in self.routes:
      self.routes.append((target, cost))

  def __lt__(self, other):
    return self.cost < other.cost

  def __str__(self):
    return self.name

  def __repr__(self):
    return f'{self._id}:{self.name}'


#Implementierung des Dijkstra-Algorithmus
def dijkstra(start, goal):
  predecessor = {start:None}              #Map mit den vorgängern um den Pfad am Ende zu rekonstrurieren
  visited = []                            #Liste mit schon abgeschlossenen Knoten
  start.cost = 0.0                        #Initiale Kosten
  q = [start]                             #Priority Queue mit den Knoten
  while q:                                #Hauptschleife
    u = heapq.heappop(q)     
    visited.append(u._id)                 #u ist jetzt abgeschlossen
    if u == goal:
      break                               #Ziel gefunden
    for i in u.routes:
      if i in visited:                    #Nachbar ist schon abgeschlossen, muss nicht nochmal untersucht werden
        continue
      if i[0] in q:                       #Nachbar ist schon in q, muss eventuell aktualisiert werden
        if u.cost + i[1] < i[0].cost:     
          i[0].cost = u.cost + i[1]
          predecessor[i[0]] = u
          heapq.heapify(q)
      elif i[0]._id not in visited:       #Nachbar noch nicht in q, wird hinzugefügt
        i[0].cost = u.cost + i[1]
        predecessor[i[0]] = u
        heapq.heappush(q, i[0])
  
  if not goal in predecessor:                       #Zielknoten ist nicht erreichbar, leerer Pfad
    return []
  
  path = []                               #Pfad rekonstruieren und zurückgeben
  p = goal
  while p != None:
    path.append(p)
    p = predecessor[p]
  
  path.reverse()
  return path


if __name__ == '__main__':
  #Suche nach der JSON Datei
  if os.path.isfile(r'generatedGraph.json'):
    p = r'generatedGraph.json'
  elif os.path.isfile(r'../data/generatedGraph.json'):
    p = r'../data/generatedGraph.json'
  else:
    print('Fehler, generatedGraph.json nicht gefunden.', file=sys.stderr)
    exit()
  
  #JSON Datei einlesen
  with open(p, 'r') as f:
    data = json.load(f)
    
  #Umwandeln der JSON Objekte in die Datenstruktur für den Dijkstra Algorithmus
  nodes = data['nodes']
  edges = data['edges']
  planets = [Planet(x, nodes[x]['label']) for x in range(len(nodes))]
  for edge in edges:
    planets[edge['source']].addRoute(planets[edge['target']], edge['cost'])
    planets[edge['target']].addRoute(planets[edge['source']], edge['cost'])
    
  #Start und Zielknoten finden    
  erde, bertrandt = None, None
  for planet in planets:
    if planet.name == 'Erde':
      erde = planet
    elif planet.name == 'b3-r7-r4nd7':
      bertrandt = planet
    if erde and bertrandt:
      break

  if erde is None or bertrandt is None:
    print('Fehler, Erde oder b3-r7-r4nd7, nicht im graph gefunden.', file=sys.stderr)
    exit()
      
  #Algorithmus ausführen und Ergebnis ausgeben
  path = dijkstra(erde, bertrandt)
  print('Lösung:')
  print(f'Gesamtentfernung: {bertrandt.cost}')
  print(f'Pfad: {" -> ".join([str(x) for x in path])}')
